Change message on the empty result of searching routes by `rails routes` with `-c` or `-g`

bogdanvlviv opened the pull request:
- Improve docs of `ActionDispatch::Routing`
  - Add a mention about `-g`.
  - Improve info about `--expanded` option of `rails routes`.

- Introduce `ActionDispatch::Routing::ConsoleFormatter::Base`
  - Create `Base` and inherit `Sheet` and `Expanded` in order to
    prevent code duplication.
    - Remove trailing "\n" for components of `Expanded`.
    - There is no need for `Expanded#header` to return `@buffer` so return `nil` instead.
  - Change `no_routes` message "No routes were found for this controller"
    since if use `-g`, it sounds incorrect.
    - Display `No routes were found for this controller.` if apply `-c`.
    - Display `No routes were found for this grep pattern.` if apply `-g`.

Related to #32130

kaspth commented on actionpack/lib/action_dispatch/routing.rb:
```diff
@@ -243,9 +243,9 @@ module ActionDispatch
-  # Target specific controllers by prefixing the command with <tt>-c</tt> option. Use
-  # <tt>--expanded</tt> to turn on the expanded table formatting mode.
-  #
+  # Target specific controller by prefixing the command with <tt>-c</tt> option,
```
Let's strike the 2 `option` occurences. The sentences sound better without them.

It should be: `Target a specific`.

eugeneius replied:
> by prefixing the command with <tt>-c</tt> option

That doesn't sound right; options are passed after the command. I'd ✂️ "by prefixing the command" here.

kaspth commented on actionpack/lib/action_dispatch/routing.rb:
```diff
@@ -243,9 +243,9 @@ module ActionDispatch
-  #
+  # Target specific controller by prefixing the command with <tt>-c</tt> option,
+  # or grep routes using <tt>-g</tt> option. Use <tt>--expanded</tt> to turn on
+  # the expanded table formatting mode.
```
`the expanded table formatting mode` doesn't really explain anything to users. How about:

```
# Useful in conjunction with <tt>--expanded</tt> which displays routes vertically.
```

kaspth commented on actionpack/lib/action_dispatch/routing/inspector.rb:
```diff
@@ -212,10 +211,10 @@ def no_routes(routes)
+            "No routes were found."
           end
           @buffer << "For more information about routes, see the Rails guide: http://guides.rubyonrails.org/routing.html."
         end
```
Seems like we should figure out a way to share the 2 `no_routes` methods.

bogdanvlviv replied:
Yeah. I'll create `Base` class and inherit from it.

kaspth commented on actionpack/lib/action_dispatch/routing/inspector.rb:
```diff
@@ -212,10 +211,10 @@ def no_routes(routes)
             MESSAGE
           else
-            "No routes were found for this controller\n"
+            "No routes were found."
```
Perhaps we should tune this to `No routes found matching conditions.` to indicate to the user that their `-g` or `-c` was applied.

bogdanvlviv replied:
@kaspth What do you think about displaying a message depending on which option was passed `-c` or `-g`.

```diff
diff --git a/actionpack/lib/action_dispatch/routing/inspector.rb b/actionpack/lib/action_dispatch/routing/inspector.rb
index 73f0531..36128b3 100644
--- a/actionpack/lib/action_dispatch/routing/inspector.rb
+++ b/actionpack/lib/action_dispatch/routing/inspector.rb
@@ -64,7 +64,7 @@ def format(formatter, filter = nil)
         routes_to_display = filter_routes(normalize_filter(filter))
         routes = collect_routes(routes_to_display)
         if routes.none?
-          formatter.no_routes(collect_routes(@routes))
+          formatter.no_routes(collect_routes(@routes), filter)
           return formatter.result
         end
 
@@ -82,9 +82,9 @@ def format(formatter, filter = nil)
       private
 
         def normalize_filter(filter)
-          if filter.is_a?(Hash) && filter[:controller]
+          if filter[:controller]
             { controller: /#{filter[:controller].downcase.sub(/_?controller\z/, '').sub('::', '/')}/ }
-          elsif filter
+          elsif filter[:grep_pattern]
             { controller: /#{filter}/, action: /#{filter}/, verb: /#{filter}/, name: /#{filter}/, path: /#{filter}/ }
           end
         end
@@ -147,7 +147,7 @@ def header(routes)
           @buffer << draw_header(routes)
         end
 
-        def no_routes(routes)
+        def no_routes(routes, filter)
           @buffer <<
           if routes.none?
             <<~MESSAGE
@@ -205,7 +205,7 @@ def section(routes)
         def header(routes)
         end
 
-        def no_routes(routes)
+        def no_routes(routes, filter)
           @buffer <<
           if routes.none?
             <<~MESSAGE
@@ -213,8 +213,10 @@ def no_routes(routes)
 
             Please add some routes in config/routes.rb.
             MESSAGE
-          else
-            "No routes were found."
+          elsif filter.has_key?(:controller)
+            "No routes were found for this controller."
+          elsif filter.has_key?(:grep_pattern)
+            "No routes were found for this pattern."
           end
           @buffer << "For more information about routes, see the Rails guide: http://guides.rubyonrails.org/routing.html."
         end
diff --git a/railties/lib/rails/commands/routes/routes_command.rb b/railties/lib/rails/commands/routes/routes_command.rb
index c4f3717..859ab9c 100644
--- a/railties/lib/rails/commands/routes/routes_command.rb
+++ b/railties/lib/rails/commands/routes/routes_command.rb
@@ -39,7 +39,7 @@ def routes_filter
           if options.has_key?("controller")
             { controller: options["controller"] }
           elsif options.has_key?("grep_pattern")
-            options["grep_pattern"]
+            { grep_pattern: options["grep_pattern"] }
           else
             nil
           end
```

bogdanvlviv committed:
Improve docs of `ActionDispatch::Routing`

- Add a mention about `-g`.
- Improve info about `--expanded` option of `rails routes`.

bogdanvlviv committed:
Introduce `ActionDispatch::Routing::ConsoleFormatter::Base`

- Create `Base` and inherit `Sheet` and `Expanded` in order to
- prevent code duplication.
  - Remove trailing "\n" for components of `Expanded`.
  - There is no need for `Expanded#header` to return `@buffer` so return `nil` instead.

- Change `no_routes` message "No routes were found for this controller"
  since if use `-g`, it sounds incorrect.
  - Display `No routes were found for this controller.` if apply `-c`.
  - Display `No routes were found for this grep pattern.` if apply `-g`.

Related to #32130

bogdanvlviv commented:
Thanks for the review! I just added suggested changes. What do you think about https://github.com/rails/rails/pull/32153#discussion_r172048705?

kaspth commented:
@bogdanvlviv 👍, if we can use `symbolize_keys.slice(:controller, :grep_pattern)` that might be better too (can't recall if we have Active Support access there).

kaspth commented:
All right, lets squash the commits then we're good 👍

bogdanvlviv commented:
Squashed commits related to changing messages on an empty result of `bin/rails routes` into one commit.
Fixing docs leave in separate commit(e78b1e5e615978a221b68530641101451ab83e68) since it isn't related  to 304906f1bd4cd6f2831b6ae39a9ae54599d85569.

kaspth commented:
Solid 👍
